Profile Page Cleanup

Allcharles opened the pull request:
# Cleanup Profile Pages

Cleaned up a few of the profile pages.

## Changes

- Created my bookmarks page
- Created boilerplate for my annotations page
- Cleanup profile page to display values
- Updated `PagedTableTemplate` class
- Updated services to find models a specific user has access to

## Issues

- Project/Site services currently cannot retrieve the possible list of user accessible models because `accessLevel` has not been added as a filterable output
- Most popular tags currently not possible to generate as they do not provide a count for the number of times they have been used. Would have to take the list of all tags, calculate how many times they have been used, and then order @atruskie 

## Closes

Relates to #137 #136 #135 #192 

## Visual Changes

### Personal Account

Profile
![image](https://user-images.githubusercontent.com/3955116/81775253-d5220700-952f-11ea-8b1e-c7f2260ea6e0.png)

Bookmarks
![image](https://user-images.githubusercontent.com/3955116/81776391-0ac7ef80-9532-11ea-9791-54f6307f670d.png)

Projects
![image](https://user-images.githubusercontent.com/3955116/81775374-161a1b80-9530-11ea-9f59-3aae6fb286f8.png)

Sites
![image](https://user-images.githubusercontent.com/3955116/81775388-1fa38380-9530-11ea-8baf-215526328601.png)

### User Account

Profile
![image](https://user-images.githubusercontent.com/3955116/81775262-da7f5180-952f-11ea-91c9-ce7abe19b561.png)

Bookmarks
![image](https://user-images.githubusercontent.com/3955116/81776402-0f8ca380-9532-11ea-8046-3a54c672ba3e.png)

Projects
![image](https://user-images.githubusercontent.com/3955116/81775443-3518ad80-9530-11ea-9125-4831a2ff1b4a.png)

Sites
![image](https://user-images.githubusercontent.com/3955116/81775456-39dd6180-9530-11ea-8c72-71a11fbfeb0f.png)

Allcharles committed:
Created bookmarks pages

Allcharles committed:
Created annotation pages

Allcharles committed:
Cleanup filter typings

Allcharles committed:
Cleanup profile site/project pages

Allcharles committed:
Cleanup profile page

Allcharles committed:
Fix bookmarks table

Allcharles committed:
Cleanup filterKeys

atruskie commented on src/app/component/admin/tag-group/list/list.component.ts:
```diff
@@ -57,6 +56,8 @@ export class AdminTagGroupsComponent extends PagedTableTemplate<
       }))
     );
+
+    this.filterKey = "groupIdentifier";
```
If this does what I think it does, probably better to be `tag.text`

Allcharles replied:
`TagGroup` does not have a text parameter. `text` is attached to the `Tag` model

atruskie commented on src/app/component/profile/pages/bookmarks/their-bookamrks.component.spec.ts:
```diff
@@ -0,0 +1,72 @@
+import { ComponentFixture, TestBed } from "@angular/core/testing";
```
filename contains spelling mistake

atruskie commented on src/app/component/shared/items/item/item.component.ts:
```diff
@@ -47,7 +47,7 @@ import { StrongRoute } from "src/app/interfaces/strongRoute";
     </div>
   `,
-  changeDetection: ChangeDetectionStrategy.OnPush
+  changeDetection: ChangeDetectionStrategy.OnPush,
```
I'd leave a comment explaining why we can use this strategy (immutable objects I presume?)

Allcharles replied:
Tbh, i should convert this to take immutable objects only (although this isn't the reason for the change detection strategy). We can set this change detection strategy because its a pure component (only thing that changes is the input so it doesn't need to do change detection). Leaving a comment would be a bit overkill but i'll add it

atruskie replied:
Well that why is exactly the context I was missing... hence the need.

atruskie commented on src/app/component/shared/items/items/items.component.ts:
```diff
@@ -38,9 +39,9 @@ import { ItemInterface } from "../item/item.component";
     </div>
   `,
-  changeDetection: ChangeDetectionStrategy.OnPush
+  changeDetection: ChangeDetectionStrategy.OnPush,
```
comment why

atruskie commented on src/app/services/baw-api/api-common.ts:
```diff
@@ -46,6 +46,30 @@ export function option(x?: New | Filter | Empty) {
+    },
+  };
+}
+
```
This method is redundant for `service<T>` and `Model<T>`. You don't need `/filter` to get a specific object of the same type... you just use `/:id`

It is useful for `service<T>` and `Model<U>` - i.e. some other model that can be queried that is associated with `T`.

Add documentation to that effect maybe?

Allcharles replied:
```typescript
/**
 * Modify a base filter to add a foreign key condition
 * @param filters Base Filters
 * @param key Foreign key
 * @param model Foreign key value
 */
export function filterByForeignId<T extends AbstractModel>(
```

atruskie commented on src/app/services/baw-api/baw-api.service.ts:
```diff
@@ -255,11 +255,88 @@ export interface Paging {
   next?: string;
 }
 
+interface Combinations {
```
Maybe `conditions`

Allcharles replied:
![image](https://user-images.githubusercontent.com/3955116/81907351-f1df3d00-960a-11ea-9b2a-87a7b4fcf50c.png)

atruskie replied:
Fair enough

atruskie commented on src/app/services/baw-api/baw-api.service.ts:
```diff
@@ -255,11 +255,88 @@ export interface Paging {
+  | string[]
+  | number[]
+  | {
+      interval?: string;
```
interval has a narrow syntax that is acceptable: `(\[|\()[-+]?[0-9]*\.?[0-9]+,[-+]?[0-9]*\.?[0-9]+(\]|\))`

May be able to strongly type it?

Allcharles replied:
Regex typing is not currently possible: https://github.com/Microsoft/TypeScript/issues/6579. Could use `Range` as type checking for an Interval class however its not recommended for performance reasons: https://github.com/microsoft/TypeScript/issues/15480
I saw that its possible to strong type the value using the following brute force "hack":
```typescript
type Interval = "[0, 1]" | "[0, 2]" | ... | ["999", "999"];
```
Depending on the size of the solution, it would be possible to use something like https://github.com/KendallHopkins/FormalTheory to generate all possible options and just store the type in a file.

Edit: Did a rough calc, and i think there are about 16 million possibilities for the regex (potentially an order of magnitude more)

atruskie replied:
Yup we can't regex match the type. Could we just make it a type though, i.e a class?

Allcharles replied:
Do you want the class to type check the input at run time, or basically just be a string wrapper?

atruskie replied:
It can do a bit of checking yeah

atruskie commented on src/app/services/baw-api/baw-api.service.ts:
```diff
@@ -255,11 +255,88 @@ export interface Paging {
+  | {
+      interval?: string;
+      from?: number;
+      to?: number;
```
`from` and `to` are exclusive to `interval`

atruskie commented on src/app/services/baw-api/baw-api.service.ts:
```diff
@@ -255,11 +255,88 @@ export interface Paging {
+      Subsets &
+      { [P in keyof T]?: Combinations & Comparisons & Subsets })
+  | {};
+
```
Wow

Allcharles replied:
This is getting close to what i want. Typing isn't perfect yet as a programmer can still add keys that aren't part of the model, and values which do not match the key. Both should be possible to fix.

Allcharles replied:
Also if i can get this working properly, it should be able to accept empty filters (`{}`). Currently i have a workaround with the `| {};` bit which im not overly happy with.

Allcharles replied:
Fixed with the following change:

```typescript
export type InnerFilter<T = {}> = Combinations &
  Comparisons &
  Subsets &
  { [P in keyof T]?: Combinations & Comparisons & Subsets };
```

I still haven't added the ability for the filter to verify the subsets are valid for the key type, however I've decided that'd be over the top.

atruskie commented on src/app/services/baw-api/bookmark/bookmarks.service.ts:
```diff
@@ -32,9 +34,18 @@ export class BookmarksService extends StandardApi<Bookmark> {
+  ): Observable<Bookmark[]> {
+    return this.apiFilter(
+      endpoint(Empty, Filter),
+      user ? filterById<Bookmark>(filters, "creatorId", user) : filters
```
reading this from GH I can't confirm, but that `creatorId` key is strongly typed? I think it is, nice!

Allcharles replied:
Yup, its strongly typed

Allcharles replied:
Only thing it doesn't do is validate that the key provided is of type Id, specifically because I'm unsure how generic I want to make the function.

atruskie reviewed (approved):
Minor nits, but on the whole a great change

Allcharles committed:
Cleanup filter typings

- Removed Range accepting `interval`, `from`, and `to` inputs inside same object
- Minimized `innerFilter`

Allcharles committed:
Cleanup filterById function

Allcharles committed:
Cleanup ItemsComponent

Allcharles committed:
Created RangeInterval class

- Updated service filter typing
- Cleanup projection typing

Allcharles committed:
Passing CI

Allcharles committed:
Merge branch 'master' into profile-bookmarks

Allcharles committed:
Fix linter
